File: generated_code/GeminiPro/Python/040_SplayTree.py
class SplayTree:
    """
    A self-contained Splay Tree class implementing a dictionary-like set for integers.
    
    This implementation includes insert, delete, and search operations, all of which
    involve the characteristic splaying step to move the accessed node (or its
    parent/successor) to the root of the tree. This optimizes future accesses
    to recently used keys.
    """

    class _Node:
        """A private inner class to represent a node in the Splay Tree."""
        def __init__(self, key, parent=None, left=None, right=None):
            self.key = key
            self.parent = parent
            self.left = left
            self.right = right

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on the subtree rooted at node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on the subtree rooted at node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splaying operation on a node, moving it to the root.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig case (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-Zig case (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:  # Zig-Zag case (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag case (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def insert(self, key):
        """
        Inserts a key into the Splay Tree.
        
        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and then the new node is splayed to the root.
        
        Args:
            key (int): The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        node = self.root
        parent = None
        while node:
            parent = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:  # Key already exists
                self._splay(node)
                return
        
        # Key does not exist, insert new node
        new_node = self._Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        self._splay(new_node)

    def search(self, key):
        """
        Searches for a key in the Splay Tree.
        
        If the key is found, the corresponding node is splayed to the root and True is returned.
        If the key is not found, the last accessed node (the would-be parent) is splayed
        to the root and False is returned.
        
        Args:
            key (int): The integer key to search for.
            
        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        node = self.root
        last_visited = self.root
        while node:
            last_visited = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:  # Key found
                self._splay(node)
                return True
        
        # Key not found, splay the last visited node
        self._splay(last_visited)
        return False

    def delete(self, key):
        """
        Deletes a key from the Splay Tree.
        
        The tree is first searched for the key, which brings the node (or its parent)
        to the root. If the key is found at the root, it is removed and the remaining
        subtrees are joined.
        
        Args:
            key (int): The integer key to delete.
        """
        if not self.root:
            return

        # Search for the key, which splays the node (or its parent) to the root
        self.search(key)

        # After search, if the root's key is not the one we want to delete,
        # the key was not in the tree.
        if self.root.key != key:
            return

        # Now, the node to delete is at the root
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # No left subtree, the right subtree becomes the new tree
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # Detach the left subtree
            left_subtree.parent = None
            
            # Find the maximum element in the left subtree
            max_in_left = left_subtree
            while max_in_left.right:
                max_in_left = max_in_left.right
            
            # Splay this maximum element to the root of the left subtree.
            # We can temporarily treat the left subtree as the main tree
            # to reuse the _splay method.
            self.root = left_subtree
            self._splay(max_in_left)
            
            # The new root of the combined tree is the splayed max_node.
            # Attach the original right subtree to its right.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/040_SplayTree.java
/**
 * A self-contained implementation of an integer set using a Splay Tree.
 * A splay tree is a self-adjusting binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 * It performs basic operations such as insertion, deletion, and search in
 * O(log n) amortized time.
 */
public class SplayTree {

    private Node root;

    /**
     * Represents a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        /**
         * Constructs a new node with the given key.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at y.
     * @param y The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. It brings the node with the given key (or the last
     * accessed node if the key is not found) to the root of the tree.
     * This is a top-down splay implementation.
     * @param key The key to splay to the root.
     */
    private void splay(int key) {
        if (root == null) {
            return;
        }

        // Use a dummy node to simplify linking of left and right subtrees.
        Node dummy = new Node(0);
        Node leftTreeMax = dummy;
        Node rightTreeMin = dummy;
        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) {
                    break;
                }
                // Zig-Zig (Right-Right rotation)
                if (key < current.left.key) {
                    current = rightRotate(current);
                    if (current.left == null) {
                        break;
                    }
                }
                // Link current node to the right subtree's left-most branch
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) {
                    break;
                }
                // Zig-Zig (Left-Left rotation)
                if (key > current.right.key) {
                    current = leftRotate(current);
                    if (current.right == null) {
                        break;
                    }
                }
                // Link current node to the left subtree's right-most branch
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else { // key == current.key
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;
        rightTreeMin.left = current.right;
        current.left = dummy.right;
        current.right = dummy.left;

        root = current;
    }

    /**
     * Searches for a key in the tree. If found, it splays the node to the root
     * and returns the key. If not found, it splays the last accessed node to the
     * root and returns null.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        splay(key);
        // After splaying, if the key is in the tree, it must be the root.
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree. If the key already exists,
     * the tree is splayed on that key, but no new node is inserted.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree on the key.
        // If key is present, it moves to the root.
        // If not, the last accessed node moves to the root.
        splay(key);

        // If key is already at the root, we are done (set behavior).
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree. If the key is not found,
     * the tree is splayed on the last accessed node, and the tree remains unchanged.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree on the key. If the key is present, it becomes the root.
        splay(key);

        // If the key is not in the tree, the root's key will not match.
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Find the maximum key in the left subtree.
            Node maxNode = leftSubtree;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }
            int maxKey = maxNode.key;

            // Splay the max key to the root of the left subtree.
            // We temporarily treat the left subtree as the whole tree for splaying.
            root = leftSubtree;
            splay(maxKey);

            // After splaying, the new root (which is the max node) has no right child.
            // We can now attach the original right subtree.
            root.right = rightSubtree;
        }
    }
}